utils/interned_string.h

```diff
@@ -0,0 +1,222 @@
+#include "mldb/compiler/compiler.h"
+
+// NOTE TO MLDB DEVELOPERS: This is an API header file.  No includes
+// should be added, especially value_description.h.
+
+#pragma once
+
+
+namespace MLDB {
+
+/*****************************************************************************/
+/* INTERNED STRING                                                           */
+/*****************************************************************************/
+
+template<size_t Bytes, typename Char = char>
+struct InternedString {
+
+    // Max Bytes is 255, as otherwise we can't determine the length
+    static_assert(Bytes < 256, "First template parameter for InternedString "
+                  "must be 255 or less");
+
+    InternedString()
+        : intLength_(0)
+    {
+    }
+
+    InternedString(const InternedString & other)
+        : InternedString()
+    {
+        append(other.data(), other.length());
+    }
+
+    template<size_t OtherBytes>
+    InternedString(const InternedString<OtherBytes, Char> & other)
+        : InternedString()
+    {
+        append(other.data(), other.length());
+    }
+
+    template<size_t OtherBytes>
+    InternedString(InternedString<OtherBytes, Char> && other) noexcept
+        : InternedString()
+    {
+        if (other.length() > Bytes) {
+            // Can't fit internally.  If the other is external, steal it
+            if (other.isExt()) {
+                intLength_ = IS_EXT;
+                extLength_ = other.extLength_;
+                extCapacity_ = other.extCapacity_;
+                extBytes_ = other.extBytes_;
+                other.intLength_ = 0;
+                return;
+            }
+        }
+
+        // Otherwise, simply append it
+        append(other.data(), other.size());
+    }
+
+    InternedString(const std::basic_string<Char> & other)
+        : InternedString()
+    {
+        append(other.data(), other.length());
+    }
+
+    InternedString & operator = (const InternedString & other)
+    {
+        InternedString newMe(other);
+        swap(newMe);
+        return *this;
+    }
+
+    InternedString & operator = (InternedString && other) noexcept
+    {
+        InternedString newMe(std::move(other));
+        swap(newMe);
+        return *this;
+    }
+
+    ~InternedString()
+    {
+        if (isExt())
+            deleteExt();
+    }
+
+    const Char * data() const
+    {
+        return isExt() ? extBytes_ : intBytes_;
+    }
+
+    void swap(InternedString & other) noexcept
+    {
+        std::swap(intLength_, other.intLength_);
+        std::swap(intBytes_[0], other.intBytes_[0]);
+        std::swap(intBytes_[1], other.intBytes_[1]);
+        std::swap(intBytes_[2], other.intBytes_[2]);
+        for (unsigned i = 0;  i < NUM_WORDS * 2 - 1;  ++i) {
+            std::swap(internalWords[i], other.internalWords[i]);
+        }
+    }
+
+    size_t size() const
+    {
+        return isExt() ? extLength_ : intLength_;
+    }
+
+    size_t length() const
```

Comment:
Since this is called from a nothrow ctor, I think this should also be marked as nothrow.

Reply:
This is a code move.  Note that the nothrow does not require functions it calls to be nothrow; it just says that if one of them does throw, the exception won't propagate outside.

Reply:
Doesn't it also make it un catchable and crash the system very badly?

Reply:
Yes!  Which is kind of the point: if something in there does throw (note that marking nothrow doesn't change at all whether something will) then it will noisily crash the program, as it's a precondition violation.  Which is exactly what you would want it to do.

Note that the reason noexcept was added was to enable things like vector resizes to be both performant and strongly exception safe; that's the reason for the annotation here.